src/Shared/NativeMethodsShared.cs

```diff
@@ -1181,19 +1236,37 @@ internal static int GetParentProcessId(int processId)
         /// Internal, optimized GetCurrentDirectory implementation that simply delegates to the native method
         /// </summary>
         /// <returns></returns>
-        internal static string GetCurrentDirectory()
+        internal unsafe static string GetCurrentDirectory()
         {
+#if FEATURE_LEGACY_GETCURRENTDIRECTORY
             if (IsWindows)
             {
-                StringBuilder sb = new StringBuilder(MAX_PATH);
-                int pathLength = GetCurrentDirectory(MAX_PATH, sb);
-
-                return pathLength > 0 ? sb.ToString() : null;
+                int bufferSize = GetCurrentDirectoryAndAssertSuccess(0, null);
+                char* buffer = stackalloc char[bufferSize];
+                int pathLength = GetCurrentDirectoryAndAssertSuccess(bufferSize, buffer);
+                return new string(buffer, startIndex: 0, length: pathLength);
             }
-
+#endif
             return Directory.GetCurrentDirectory();
         }
 
+        private unsafe static int GetCurrentDirectoryAndAssertSuccess(int nBufferLength, char* lpBuffer)
+        {
+            int pathLength = GetCurrentDirectory(nBufferLength, lpBuffer);
+            AssertNativeMethodSuccess(pathLength);
+            return pathLength;
+        }
+
+        internal static void AssertNativeMethodSuccess(int result)
```

Comment:
Following our [ErrorUtilities convention](https://github.com/Microsoft/msbuild/blob/7f95dc1575d19b9fac4da14b6f754aefe9d73a12/src/Shared/ErrorUtilities.cs#L617), I'd probably call this `VerifyThrowWin32Result(int result)`